src/atsam/fdcan.c

```diff
@@ -0,0 +1,309 @@
+void
+CAN_IRQHandler(void)
+{
+    uint32_t ir = CANx->MCAN_IR;
+
+    if (ir & MCAN_IE_RF0NE) {
+        CANx->MCAN_IR = MCAN_IE_RF0NE;
+
+        uint32_t rxf0s = CANx->MCAN_RXF0S;
+        if (rxf0s & MCAN_RXF0S_F0FL_Msk) {
+            // Read and ack data packet
+            uint32_t idx = (rxf0s & MCAN_RXF0S_F0GI_Msk) >> MCAN_RXF0S_F0GI_Pos;
+            struct fdcan_fifo *rxf0 = &MSG_RAM.RXF0[idx];
+            uint32_t ids = rxf0->id_section;
+            struct canbus_msg msg;
+            if (ids & FDCAN_XTD)
+                msg.id = (ids & 0x1fffffff) | CANMSG_ID_EFF;
+            else
+                msg.id = (ids >> 18) & 0x7ff;
+            msg.id |= ids & FDCAN_RTR ? CANMSG_ID_RTR : 0;
+            msg.dlc = (rxf0->dlc_section >> 16) & 0x0f;
+            msg.data32[0] = rxf0->data[0];
+            msg.data32[1] = rxf0->data[1];
+            barrier();
+            CANx->MCAN_RXF0A = idx;
+
+            // Process packet
+            canbus_process_data(&msg);
+        }
+    }
+    if (ir & FDCAN_IE_TC) {
+        // Tx
+        CANx->MCAN_IR = FDCAN_IE_TC;
+        canbus_notify_tx();
+    }
+}
+
+static inline const uint32_t
+make_btr(uint32_t sjw,       // Sync jump width, ... hmm
+         uint32_t time_seg1, // time segment before sample point, 1 .. 16
+         uint32_t time_seg2, // time segment after sample point, 1 .. 8
+         uint32_t brp)       // Baud rate prescaler, 1 .. 1024
+{
+    return (((uint32_t)(sjw-1)) << MCAN_NBTP_NSJW_Pos
+            | ((uint32_t)(time_seg1-1)) << MCAN_NBTP_NTSEG1_Pos
+            | ((uint32_t)(time_seg2-1)) << MCAN_NBTP_NTSEG2_Pos
+            | ((uint32_t)(brp - 1)) << MCAN_NBTP_NBRP_Pos);
+}
+
+static inline const uint32_t
+compute_btr(uint32_t pclock, uint32_t bitrate)
+{
+    /*
+        Some equations:
+        Tpclock = 1 / pclock
+        Tq      = brp * Tpclock
+        Tbs1    = Tq * TS1
+        Tbs2    = Tq * TS2
+        NominalBitTime = Tq + Tbs1 + Tbs2
+        BaudRate = 1/NominalBitTime
+        Bit value sample point is after Tq+Tbs1. Ideal sample point
+        is at 87.5% of NominalBitTime
+        Use the lowest brp where ts1 and ts2 are in valid range
+     */
+
+    uint32_t bit_clocks = pclock / bitrate; // clock ticks per bit
+
+    uint32_t sjw = 2;
+    uint32_t qs;
+    // Find number of time quantas that gives us the exact wanted bit time
+    for (qs = 18; qs > 9; qs--) {
+        // check that bit_clocks / quantas is an integer
+        uint32_t brp_rem = bit_clocks % qs;
+        if (brp_rem == 0)
+            break;
+    }
+    uint32_t brp       = bit_clocks / qs;
+    uint32_t time_seg2 = qs / 8; // sample at ~87.5%
+    uint32_t time_seg1 = qs - (1 + time_seg2);
+
+    return make_btr(sjw, time_seg1, time_seg2, brp);
+}
+
+void
+can_init(void)
+{
+#if !CONFIG_USBCANBUS
+    // Configure UPLL for USB and CAN
+    PMC->CKGR_UCKR = CKGR_UCKR_UPLLCOUNT(3) | CKGR_UCKR_UPLLEN;
+    while ( !(PMC->PMC_SR & PMC_SR_LOCKU) )
+    {
+    }
+#endif
```

Comment:
It is preferable to avoid `#ifdefs` where possible.  It would be preferable to write this as:
```
    if (!CONFIG_USBCANBUS) {
        ...
    }
```

Also, please follow the syntax conventions elsewhere in this file and use:
```
    while (...)
        ;
```